src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp

```diff
@@ -83,37 +123,317 @@ namespace
+                camera_vertex.m_bsdf->evaluate(
+                    camera_vertex.m_bsdf_data,
+                    false,   // Adjoint
+                    false,
+                    static_cast<Vector3f>(camera_vertex.m_geometric_normal),
+                    camera_vertex.m_shading_basis,
+                    static_cast<Vector3f>(foundation::normalize(light_vertex.m_position - camera_vertex.m_position)),
+                    static_cast<Vector3f>(camera_vertex.m_dir_to_prev_vertex),
+                    ScatteringMode::All,
+                    camera_eval_bsdf);
+
+                DirectShadingComponents light_eval_bsdf;
+                light_vertex.m_bsdf->evaluate(
+                    light_vertex.m_bsdf_data,
+                    true,   // Adjoint
+                    false,
+                    static_cast<Vector3f>(light_vertex.m_geometric_normal),
+                    light_vertex.m_shading_basis,
+                    static_cast<Vector3f>(foundation::normalize(camera_vertex.m_position - light_vertex.m_position)),
+                    static_cast<Vector3f>(light_vertex.m_dir_to_prev_vertex),
+                    ScatteringMode::All,
+                    light_eval_bsdf);
+
+                Spectrum geometry = ComputeGeometryTerm(shading_context, shading_point, camera_vertex, light_vertex);
+                result = geometry * camera_eval_bsdf.m_beauty * light_eval_bsdf.m_beauty * camera_vertex.m_beta * light_vertex.m_beta;
+            }
+
+            // check if connection is near black or not
+            bool nearBlack = true;
+            for (size_t i = 0; i < result.size(); i++)
+            {
+                // need nearBlack() function for spectrum
+                if (result[i] > 1e-9)
+                {
+                    nearBlack = false;
+                    break;
+                }
+            }
+
+            if (nearBlack) { return; }
+
+            float numerator = ComputePathDensity(light_vertices, camera_vertices, s, t, s, t);
+            float denominator = 0.0;
+
+            // [p = 0, q = s + t], [p = 1, q = s + t - 1] ... [p = s + t - 2, q = 2]
+            for (int i = 0; i <= s + t - 2; i++)
+            {
+                denominator += ComputePathDensity(light_vertices, camera_vertices, s, t, i, s + t - i);
+            }
+
+            float miWeight = numerator / denominator;
+
+            assert(miWeight <= 1.0f);
+            radiance.m_beauty += miWeight * result;
+        }
+
         void compute_lighting(
             SamplingContext&        sampling_context,
             const PixelContext&     pixel_context,
             const ShadingContext&   shading_context,
             const ShadingPoint&     shading_point,
             ShadingComponents&      radiance) override      // output radiance, in W.sr^-1.m^-2
         {
-            PathVisitor light_path_visitor;
-            VolumeVisitor volume_visitor;
+            std::vector<BDPTVertex> camera_vertices;
```

Comment:
We can't use `std::vector<>` during rendering as that causes memory allocations. You could use dedicated arenas for instance.